viewer/packages/rendering/src/glsl/sector/primitives/generalCylinder.vert

```diff
@@ -67,10 +72,29 @@ void main() {
     vec3 left = normalize(cross(objectToCameraModelSpace, lDir));
     vec3 up = normalize(cross(left, lDir));
 
-    vec3 localBillboardPosition = center + mat3(halfHeight * lDir, a_radius * left, a_radius * up) * position;
+    mat3 billboardWorldRotation = mat3(lDir, left, up);
+    vec3 cylinderAxisScales = vec3(halfHeight, a_radius, a_radius);
+    mat3 inverseBillboardWorldRotation = transpose(billboardWorldRotation);
+    vec3 cameraPosInCylinderSpace = inverseBillboardWorldRotation * (rayOrigin - center);
+
+    mat3 billboardWorldScaleRotation = mat3(halfHeight * lDir, a_radius * left, a_radius * up);
+
+    vec3 localBillboardPosition = center + billboardWorldScaleRotation * position;
     vec3 viewBillboardPosition = mul3(modelToView, localBillboardPosition);
 
-    gl_Position = projectionMatrix * vec4(viewBillboardPosition, 1.0 );
+    float near = projectionMatrix[3][2] / (projectionMatrix[2][2] - 1.0);
+
+    // Check whether we are inside the primitive, in which case the quad must cover the entire screen
+    if (isWithinSpan(cameraPosInCylinderSpace, cylinderAxisScales + vec3(near))) {
+        float tanFov = 1.0 / projectionMatrix[1][1];
+
+        float aspect = projectionMatrix[1][1] / projectionMatrix[0][0];
+        float maxAspect = max(aspect, 1.0 / aspect);
+        vec3 fullScreenQuadCorner = vec3(position.xy * maxAspect * tanFov * near, -near - 1e-6);
+        viewBillboardPosition = fullScreenQuadCorner;
+    }
```

Comment:
Try making sure at most one fragment is computed per pixel.

Reply:
In hinsight, I see that this already ensures only one of the quads cover the screen. The other quad stretches into the z-direction in the yz-plane (the vertex positions are defined in `primitiveGeometries.ts`, `setConeGeometry()`). Thus, it will be projected to a line at the top of the screen and effectively ignored.